### PR TITLE
Button should respect polymorphic as prop

### PR DESCRIPTION
**Pull request by mattcosta7**
Describe your changes here.

Closes https://github.com/primer/react/issues/2118

It seems like `LinkButton` was removed under the expectation that `Button` would act in apolymorphic way and handle styling links via the `As` prop, however it seems around the same time that functionality was blocked, since any `as` prop passed to button is overridden by the default prop here [see details](https://github.com/primer/react/pull/1880)

I've updated the types to account for this, and updated the test to avoid skipping the `as` prop component test.

This should allow projects to avoid using the now undocumented `LinkButton` and instead allow for properly passing as props to `Button` directly

### Screenshots

n/a

### Merge checklist

- [x] Added/updated tests
- [x] Added/updated documentation
- [ ] Tested in Chrome
- [ ] Tested in Firefox
- [ ] Tested in Safari
- [ ] Tested in Edge

Take a look at the [What we look for in reviews](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#what-we-look-for-in-reviews) section of the contributing guidelines for more information on how we review PRs.

### CODE REVIEW

**Comment by mattcosta7** [comment]
Will take a look at ci here a bit later, if the changes seems reasonable (or at least the spirit of this change)

File: package-lock.json [hunk 1]
```diff
@@ -1,12 +1,12 @@
 {
   "name": "@primer/react",
-  "version": "35.3.0",
```
Comment (mattcosta7) [inline package-lock.json:3]: looks like publishing the last version didn't bump this

File: src/Button/Button.tsx [hunk 2]
```diff
@@ -5,7 +5,7 @@ import {ButtonBase} from './ButtonBase'
 const ButtonComponent = forwardRef<HTMLButtonElement, ButtonProps>(
   ({children, ...props}, forwardedRef): JSX.Element => {
     return (
-      <ButtonBase ref={forwardedRef} {...props} as="button">
+      <ButtonBase ref={forwardedRef} as="button" {...props}>
```
Comment (mattcosta7) [inline src/Button/Button.tsx:8]: allow `as` to be overwritten by passed in props

File: src/Button/types.ts [hunk 3]
```diff
@@ -13,7 +13,10 @@ export type VariantType = 'default' | 'primary' | 'invisible' | 'danger' | 'outl
 
 export type Size = 'small' | 'medium' | 'large'
 
-type StyledButtonProps = ComponentPropsWithRef<typeof StyledButton>
+/**
+ * Remove styled-components polymorphic as prop, which conflicts with radix's
+ */
+type StyledButtonProps = Omit<ComponentPropsWithRef<typeof StyledButton>, 'as'>
```
Comment (mattcosta7) [inline src/Button/types.ts:19]: the crux of the fix here seems like the `as` prop applied by styled-components differs from the types of that applied using the Radix polymorphic helper types an an incompatible way.  By omitting the styled component version of this type from the types that get passed around, we can rely on the radix version and avoid weird issues when trying to apply the as prop to the button base to the left of the spread restProps

**Review by jdrush89 — APPROVED** [review]

**Review by hebertcisco — APPROVED** [review]

**Review by pksjce — APPROVED** [review]
Added back a few stories using this feature. LGTM 👍

**Pull request merged**